File: elixir/day1/lib/part2.ex
defmodule Part2 do
  def parse_line(line) do
    line
    |> String.split(~r/\s+/, trim: true)
    |> then(fn [left_str, right_str] ->
      {String.to_integer(left_str), String.to_integer(right_str)}
    end)
  end

  def parse_input(input) do
    {lefts, rights} =
      input
      |> String.split("\n", trim: true)
      |> Enum.map(&parse_line/1)
      |> Enum.unzip()

    {Enum.sort(lefts), Enum.sort(rights)}
  end

  def part2(input) do
    {left_list, right_list} = input |> parse_input()

    right_map = Enum.frequencies(right_list)

    Enum.reduce(left_list, 0, fn left, acc ->
      case Map.get(right_map, left) do
        nil -> acc
        count -> acc + left * count
      end
    end)
  end
end

File.read!("input.txt")
|> Part2.part2()
|> IO.inspect(label: "part2")

File: ruby/day1/part2.rb
# frozen_string_literal: true

def parse_input(input)
  left, right = input.split("\n")
                     .map { |line| line.split.map(&:to_i) }
                     .transpose
  [left.sort, right.sort]
end

def part2(input)
  left_list, right_list = parse_input(input)

  right_map = right_list.tally

  left_list.reduce(0) do |acc, left|
    count = right_map[left]
    count.nil? ? acc : acc + (left * count)
  end
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  p part2(input)
end
